### PR TITLE
txt.py restructuring and acq read addition

### PR DESCRIPTION
**Pull request by vinferrer**
Closes issue #137 and #136. 
Add acq support mentioned in issue #41
## Proposed Changes

 - restructuring ``populate_phys_input`` function so it only checks the header and recognizes the file type. File reading has been delegated to secondary functions. Added commented files support both in header and channels 
 - process_labchart: new function that process labchart header and channels
 - process_acq: new function that process header and channels
 - read_header_and_channels: new function that reads txt files and separates them into header and channels 
- still missing multi-frequency support

### CODE REVIEW

File: phys2bids/interfaces/txt.py [hunk 1]
```diff
@@ -22,21 +22,56 @@ def populate_phys_input(filename, chtrig):
...
+                line.remove('')
             try:
                 float(line[0])
             except ValueError:
                 header.append(line)
                 continue
             line = [float(i) for i in line]
             channel_list.append(line)
+        if len(header) == 0:
+            raise AttributeError(f'Files without header are not supported yet')
```
Comment (eurunuela) [inline phys2bids/interfaces/txt.py]: `raise AttributeError('Files without header are not supported yet')`

Comment (eurunuela) [inline phys2bids/interfaces/txt.py]:
```suggestion
            raise AttributeError('Files without header are not supported yet')
```

File: phys2bids/interfaces/txt.py [hunk 2]
```diff
@@ -22,21 +22,56 @@ def populate_phys_input(filename, chtrig):
...
+        elif 'Interval=' in header[0]:
+            print('phys2bids detected that your file is in labchart format')
+            phys_in = labchart_read(channel_list, chtrig, header)
+        elif 'acq' in header[0][0]:
+            phys_in = acq_read(channel_list, chtrig, header)
+        else:
+            raise AttributeError(f'This file format is not supported yet for txt files')
```
Comment (eurunuela) [inline phys2bids/interfaces/txt.py]:
```suggestion
            raise AttributeError('This file format is not supported yet for txt files')
```

File: phys2bids/interfaces/txt.py [hunk 3]
```diff
@@ -22,21 +22,56 @@ def populate_phys_input(filename, chtrig):
...
+        index of trigger channel
+    header: list
+        list with that contains file header
+    Output
+    ------------------
+    BlueprintInput object for more see BlueprintInput docs
+    """
     # get frequency
+    if len(header) == 0:
+        raise AttributeError(f'Files without header are not supported yet')
```
Comment (eurunuela) [inline phys2bids/interfaces/txt.py]:
```suggestion
        raise AttributeError('Files without header are not supported yet')
```

File: phys2bids/interfaces/txt.py [hunk 4]
```diff
@@ -77,3 +115,70 @@ def populate_phys_input(filename, chtrig):
     timeseries.pop(0)
     ordered_timeseries = ordered_timeseries + timeseries
     return BlueprintInput(ordered_timeseries, freq, names, units)
+
+
+def acq_read(channel_list, chtrig, header=[]):
+    """
+    reading function for acq files in txt format
```
Comment (eurunuela) [inline phys2bids/interfaces/txt.py]:
```suggestion
    Reading function for acq files in txt format
```

File: phys2bids/interfaces/txt.py [hunk 5]
```diff
@@ -77,3 +115,70 @@ def populate_phys_input(filename, chtrig):
...
+    ------------------
+    channel_list: list
+        list with channels only
+    chtrig : int
+        index of trigger channel
+    header: list
+        list with that contains file header
+    Output
+    ------------------
+    BlueprintInput object for more see BlueprintInput docs
```
Comment (eurunuela) [inline phys2bids/interfaces/txt.py]:
```suggestion
    BlueprintInput object
```

File: phys2bids/interfaces/txt.py [hunk 6]
```diff
@@ -77,3 +115,70 @@ def populate_phys_input(filename, chtrig):
...
+    """
+    # get frequency
+    if len(header) == 0:
+        raise AttributeError(f'Files without header are not supported yet')
```
Comment (eurunuela) [inline phys2bids/interfaces/txt.py]:
```suggestion
        raise AttributeError('Files without header are not supported yet')
```

File: phys2bids/interfaces/txt.py [hunk 7]
```diff
@@ -77,3 +115,70 @@ def populate_phys_input(filename, chtrig):
...
+        elif interval[-1] == 'µs':
+            interval[0] = float(interval[0]) / 1000000
+            interval[-1] = 's'
+    else:
+        interval[0] = float(interval[0])
+        interval[-1] = 's'
+    # get units and names
+    orig_units = []
+    orig_names = []
+    for kj in range(3, len(header[-1]) + 8, 2):
```
Comment (eurunuela) [inline phys2bids/interfaces/txt.py]: Could you use a more descriptive name for `kj`?

File: phys2bids/interfaces/txt.py [hunk 8]
```diff
@@ -22,21 +22,56 @@ def populate_phys_input(filename, chtrig):
...
+    phys_in: BlueprintInput object for more see BlueprintInput docs
     """
+
     header = []
     channel_list = []
     with open(filename, 'r') as f:
         for line in f:
             line = line.rstrip('\n').split('\t')
+            for item in line:
+                if '#' in item:  # detecting comments
```
[the remark this answers is not in the document]
Comment (vinferrer) [reply phys2bids/interfaces/txt.py]: Well I could change the condition to:
```suggestion
                if '#' == item[0]:  # detecting comments
```
This way it should avoid these kind of problems

File: phys2bids/interfaces/txt.py [hunk 9]
```diff
@@ -22,21 +22,56 @@ def populate_phys_input(filename, chtrig):
...
+            print('phys2bids detected that your file is in labchart format')
+            phys_in = labchart_read(channel_list, chtrig, header)
+        elif 'acq' in header[0][0]:
+            phys_in = acq_read(channel_list, chtrig, header)
+        else:
+            raise AttributeError(f'This file format is not supported yet for txt files')
+    return phys_in
+
+
+def labchart_read(channel_list, chtrig, header=[]):
```
[the remark this answers is not in the document]
Comment (vinferrer) [reply phys2bids/interfaces/txt.py]: I think it doesn't matter from a functional point of view, I tried the phys2bids command before doing the pull request and is working, for the ``Test_belt_samefreq.txt``, actually the plot is the same for ``Test_belt_samefreq.acq`` so there shouldn't be any problem. However you might be right that is better to mention first functions that are used in the main function, like if was a main.py

[the remark this answers is not in the document]
Comment (vinferrer) [reply phys2bids/interfaces/txt.py]: done

File: phys2bids/interfaces/txt.py [hunk 10]
```diff
@@ -77,3 +73,121 @@ def populate_phys_input(filename, chtrig):
...
+    units = ['s', 'Volts']
+    orig_units.pop(chtrig - 1)
+    units = units + orig_units
+    # get channels
+    timeseries = np.matrix(channel_list).T.tolist()
+    freq = [1 / interval[0]] * len(timeseries)
+    timeseries = [np.array(darray) for darray in timeseries]
+    duration = (timeseries[0].shape[0] + 1) * interval[0]
+    t_ch = np.ogrid[0:duration:interval[0]][:-1]  # create time channel
+    # step not equal to sample rate, check with stephano
```
Comment (smoia) [inline phys2bids/interfaces/txt.py]:
```suggestion
    # step not equal to sample rate, check with Stefano
```

File: phys2bids/interfaces/txt.py [hunk 11]
```diff
@@ -77,3 +73,121 @@ def populate_phys_input(filename, chtrig):
...
+        elif interval[-1] == 'µs':
+            interval[0] = float(interval[0]) / 1000000
+            interval[-1] = 's'
+    else:
+        interval[0] = float(interval[0])
+        interval[-1] = 's'
+    # get units and names
+    orig_units = []
+    orig_names = []
+    for index1 in range(3, len(header[-1]) + 8, 2):
```
Comment (smoia) [inline phys2bids/interfaces/txt.py]: I think that @eurunuela 's point is still not met. Can you maybe comment this line to make it more clear (so we can help you with a name for the variable?)

File: phys2bids/interfaces/txt.py [hunk 12]
```diff
@@ -9,34 +9,27 @@
...
-                header.append(line)
-                continue
-            line = [float(i) for i in line]
-            channel_list.append(line)
     # get frequency
+    if len(header) == 0:
+        raise AttributeError('Files without header are not supported yet')
     interval = header[0][1].split(" ")
     if interval[-1] not in ['hr', 'min', 's', 'ms', 'µs']:
         raise AttributeError(f'Interval unit "{interval[-1]}" is not in a valid LabChart'
```
Comment (smoia) [inline phys2bids/interfaces/txt.py]: Can you add a comment with the reference of the valid LabChart units (if it exists)?

File: phys2bids/interfaces/txt.py [hunk 13]
```diff
@@ -9,34 +9,27 @@
...
+def labchart_read(channel_list, chtrig, header=[]):
     """
-    Populate object phys_input
-    for now this works only with labchart files
-        Input (Properties)
-    ------------------
-    filename: str
-        path to the txt labchart file
+    Reading function for labchart files
+    Parameters
```
Comment (smoia) [inline phys2bids/interfaces/txt.py]: Valid for all docstrings:
I'm doubting that an empty line is needed before the "title" (e.g. Parameters). At least, I see numpy functions having them:
https://github.com/numpy/numpy/blob/d9b1e32cb8ef90d6b4a47853241db2a28146a57d/numpy/lib/function_base.py#L1284
Did you try to import this function in Spyder and read its help? Or make an html page with sphinx?
Please check that it's ok to not have an empty line before all the titles.

Also, please add **Raises** sections (https://numpydoc.readthedocs.io/en/latest/format.html), please?

File: phys2bids/interfaces/txt.py [hunk 14]
```diff
@@ -77,3 +73,121 @@ def populate_phys_input(filename, chtrig):
...
+    See Also
+    --------
+    physio_obj.BlueprintInput
+    """
+    # get frequency
+    if len(header) == 0:
+        raise AttributeError('Files without header are not supported yet')
+    interval = header[1][0].split()
+    if interval[-1].split('/')[0] not in ['hr', 'min', 'sec', 'µsec', 'msec']:
+        raise AttributeError(f'Interval unit "{interval[-1]}" is not in a valid LabChart'
```
Comment (smoia) [inline phys2bids/interfaces/txt.py]: At the very least, the name of the software should change ;)
Can you put a reference for this point, if it exists?

File: phys2bids/interfaces/txt.py [hunk 15]
```diff
@@ -77,3 +73,121 @@ def populate_phys_input(filename, chtrig):
...
+                float(line[0])
+            except ValueError:
+                header.append(line)
+                continue
+            line = [float(i) for i in line]
+            channel_list.append(line)
+        if len(header) == 0:
+            raise AttributeError('Files without header are not supported yet')
+        elif 'Interval=' in header[0]:
+            print('phys2bids detected that your file is in labchart format')
```
Comment (smoia) [inline phys2bids/interfaces/txt.py]: You have a nice message for labchart, why not having it for AcqKnowledge too?

File: phys2bids/interfaces/txt.py (hunk 11 above)
Comment (vinferrer) [reply phys2bids/interfaces/txt.py]: so basically the variable is an index than used into the header retreats for every channel the original name and units

Comment (vinferrer) [reply phys2bids/interfaces/txt.py]: we could call it ``retriving_indx```?

File: phys2bids/interfaces/txt.py (hunk 12 above)
Comment (vinferrer) [reply phys2bids/interfaces/txt.py]: Added in the docstring

File: phys2bids/interfaces/txt.py (hunk 14 above)
Comment (vinferrer) [reply phys2bids/interfaces/txt.py]: In the case of ``acq`` I couldn't find the proper units for interval, maybe people from Chicago can help?

Comment (smoia) [reply phys2bids/interfaces/txt.py]: They don't use AcqKnowledge - maybe someone else can. For the moment, I would not raise an error in this case - but you can raise a warning!

Comment (eurunuela) [reply phys2bids/interfaces/txt.py]: I don't think µsec will be an option... Specially because it's a special character.

Comment (vinferrer) [reply phys2bids/interfaces/txt.py]: Actually, it's an option in labchart, why not in AcqKnowlegde?

File: phys2bids/interfaces/txt.py [hunk 16]
```diff
@@ -9,48 +9,46 @@
 from phys2bids.physio_obj import BlueprintInput
 
 
-def populate_phys_input(filename, chtrig):
+def labchart_read(channel_list, chtrig, header=[]):
```
Comment (eurunuela) [inline phys2bids/interfaces/txt.py]:
```suggestion
def read_labchart(channel_list, chtrig, header=[]):
```

File: phys2bids/interfaces/txt.py [hunk 17]
```diff
@@ -82,3 +83,164 @@ def populate_phys_input(filename, chtrig):
...
+    return BlueprintInput(ordered_timeseries, freq, names, units)
+
+
+def read_header_and_channels(filename, chtrig):
+    header = []
+    channel_list = []
+    with open(filename, 'r') as f:
+        for line in f:
+            line = line.rstrip('\n').split('\t')
+            if line[-1] == '':
```
Comment (eurunuela) [inline phys2bids/interfaces/txt.py]:
```suggestion
            if line[-1] == ' ':
```

Comment (eurunuela) [reply phys2bids/interfaces/txt.py]: Shouldn't it be like this if there's a final space?

File: phys2bids/tests/test_txt.py [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+    assert phys_obj.freq[0] == 1000
+    # test when units are in MHz:
+    header[1][0] = '1 MHz'
+    phys_obj = txt.acq_read(channels, chtrig, header)
+    assert phys_obj.freq[0] == 1000000
+    # test when units are not valid:
+    header[1][0] = '1 GHz'
+    with raises(AttributeError) as errorinfo:
+        phys_obj = txt.acq_read(channels, chtrig, header)
+    assert 'not in a valid AcqKnowledge' in str(errorinfo.value)
```
Comment (eurunuela) [inline phys2bids/tests/test_txt.py]: There is an empty line missing at the end of the file that will raise a linting error.

File: phys2bids/tests/test_txt.py [hunk 19]
```diff
@@ -0,0 +1,120 @@
...
+    phys_obj = txt.populate_phys_input(test_full_path, chtrig)
+    output = sys.stdout.getvalue()
+    sys.stdout = stdout
+    # check the printing output according to each format
+    output = output.split('\n')
+    assert 'AcqKnowledge format' in output[0]
+    assert 'Labchart format' in output[2]
+
+
+def test_labchart_read():
```
Comment (eurunuela) [inline phys2bids/tests/test_txt.py]:
```suggestion
def test_read_labchart():
```

File: phys2bids/interfaces/txt.py (hunk 17 above)
Comment (vinferrer) [reply phys2bids/interfaces/txt.py]: No, what happens is that due to the extra ``/t`` there is an extra field in line that is blank so I just eliminate it

File: phys2bids/interfaces/txt.py (hunk 16 above)
Comment (vinferrer) [reply phys2bids/interfaces/txt.py]: You want me to change the function name?

**Comment by vinferrer** [comment]
@smoia please review

**Comment by vinferrer** [comment]
Anddddd Travis is working now, I think you can accept the pull request safely @smoia, @eurunuela

**Comment by vinferrer** [comment]
Changes requested done, please accept the review, @smoia @eurunuela

File: phys2bids/interfaces/txt.py (hunk 16 above)
Comment (vinferrer) [reply phys2bids/interfaces/txt.py]: names changed

File: phys2bids/tests/data/Test_belt_pulse_samefreq.txt [hunk 20]
```diff
@@ -20,7 +20,7 @@ CH3	CH4	CH5	CH6	CH10	CH11
 -0.128784	-0.290527	0.0012207	-0.0302124	0.000610352	0.000305176	
 -0.129089	-0.290527	0.000305176	-0.0305176	0.0012207	0.000305176	
 -0.128784	-0.290527	0.0012207	-0.0305176	0.000915527	0	
--0.128784	-0.291138	0.0012207	-0.0302124	0.000610352	0.000305176	
+-0.128784	-0.291138	0.0012207	-0.0302124	0.000610352	0.000305176	# add coment
```
[the remark this answers is not in the document]
Comment (vinferrer) [reply phys2bids/tests/data/Test_belt_pulse_samefreq.txt]: File restored to normal and new file added that already had a comment on it

**Pull request merged**